stub/src/main/java/io/grpc/stub/BlockingClientCall.java

```diff
@@ -0,0 +1,403 @@
+   * @throws TimeoutException if no read becomes ready before the specified timeout expires
+   * @throws io.grpc.StatusRuntimeException If the stream has closed in an error state
+   */
+  public RespT read(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException {
+    return read(false, timeout, unit);
+  }
+
+  private RespT read(boolean waitForever, long timeout, TimeUnit unit)
+      throws InterruptedException, TimeoutException {
+    long start = System.nanoTime();
+    long end = start + unit.toNanos(timeout);
+
+    executor.drain();
+
+    RespT bufferedValue;
+    while ((bufferedValue = buffer.poll()) == null && closedStatus == null) {
+      Predicate<BlockingClientCall<ReqT, RespT>> predicate = BlockingClientCall::skipWaitingForRead;
+      waitAndDrainExecutorOrTimeout(waitForever, end, predicate, this);
+    }
+
+    if (logger.isLoggable(Level.FINER)) {
+      logger.finer("Client Blocking read had value:  " + bufferedValue);
+    }
+
+    if (bufferedValue != null) {
+      call.request(1);
+    } else if (closedStatus != null && !closedStatus.isOk()) {
+      throw closedStatus.asRuntimeException();
+    }
+
+    return bufferedValue;
+  }
+
+  boolean skipWaitingForRead() {
+    return closedStatus != null || !buffer.isEmpty();
+  }
+
+  /**
+   * Wait for a value to be available from the server. If there is an
+   * available value, return true immediately.  If the stream was closed with Status.OK, return
+   * false.  If the stream was closed with an error status, throw a StatusException. Otherwise, wait
+   * for a value to be available or the stream to be closed.
+   *
+   * @return True when there is a value to read.  Return false if stream closed cleanly.
+   * @throws io.grpc.StatusRuntimeException If the stream was closed in an error state
+   */
+  public boolean hasNext() throws InterruptedException {
+    executor.drain();
+
+    if (!buffer.isEmpty()) {
+      return true;
+    }
+
+    if (closedStatus != null) {
+      if (!closedStatus.isOk()) {
+        throw closedStatus.asRuntimeException();
+      }
+      return false;
+    }
+
+    while (!isReadReady() && closedStatus == null) {
+      executor.waitAndDrain();
+    }
+
+
+    if (closedStatus != null && !closedStatus.isOk()) {
+      throw closedStatus.asRuntimeException();
+    }
+
+    return !buffer.isEmpty();
+  }
+
+  /**
+   * Send a value to the stream for sending to server, wait if necessary for the grpc stream to be
+   * ready.
+   * <br>
+   * If write is not legal at the time of call, immediately returns false
+   * <br><br>
+   * <b>NOTE:  </b>This method will return as soon as it passes the request to the grpc stream
+   * layer. It will not block while the message is being sent on the wire and returning true does
+   * not guarantee that the server gets the message.
+   * <p>
+   * <b>WARNING:  </b>Doing only writes without reads can lead to deadlocks as a result of flow
```

Comment:
"flow control between the client and server."? To try and make clear this is a network thing, and not an API issue.

Reply:
done